### PR TITLE
[search] Implemented viewport search results filtering.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov @rokuz PTAL

### CODE REVIEW

File: map/framework.cpp [hunk 1]
```diff
@@ -1349,6 +1350,15 @@ bool Framework::Search(search::SearchParams const & params)
 
   // Cancels previous search request (if any) and initiates new search request.
   CancelQuery(intent.m_handle);
+
+  {
+    m2::PointD const defaultMarkSize = GetSearchMarkSize(SearchMarkType::DefaultSearchMark);
+    m2::PointD const bookingMarkSize = GetSearchMarkSize(SearchMarkType::BookingSearchMark);
+    double const eps = std::max(std::max(defaultMarkSize.x, defaultMarkSize.y),
```
Comment (mpimenov) [inline map/framework.cpp]: No ```std::```, please.

File: search/nearby_points_sweeper.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#include "search/nearby_points_sweeper.hpp"
+
+namespace search
+{
+// NearbyPointsSweeper::Event ----------------------------------------------------------------------
+NearbyPointsSweeper::Event::Event(Type type, double y, double x, size_t index)
```
Comment (mpimenov) [inline search/nearby_points_sweeper.cpp]: Could you please interchange ```x``` and ```y``` here?

File: search/nearby_points_sweeper.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#include "std/algorithm.hpp"
+#include "std/cstdint.hpp"
+#include "std/limits.hpp"
+#include "std/set.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+namespace search
+{
+// This class can be used to sweep too close points on a plane.  Two
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: Лучше "points on a plane that are too close", то тоже не супер.

File: search/nearby_points_sweeper.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+// points are considered to be "too close" when Manhattan distance
+// between them is less than or equal to some pre-selected epsilon.
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: The hyphen is superfluous here.

File: search/nearby_points_sweeper.hpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+public:
+  explicit NearbyPointsSweeper(double eps);
+
+  // Adds a new point (|x|, |y|) on the plane. |index| is used to
+  // identify individual points, and will be reported for survived
+  // points during the Sweep phase.
+  void Add(double x, double y, size_t index);
+
+  // Emits indexes of all survived points. Complexity: O(n * log(n)),
+  // where n is a number of already added points.
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: s/a number/the number/

File: search/pre_ranker.cpp [hunk 6]
```diff
@@ -45,6 +45,40 @@ struct ComparePreResult1
...
+  {
+    size_t const j = rng() % (i + 1);
+    if (j < k)
+      result[j] = i;
+  }
+
+  return result;
+}
+
+void SweepNearbyResults(double eps, vector<PreResult1> & results)
```
Comment (mpimenov) [inline search/pre_ranker.cpp]: Up to you, but I'd change the variable name to something neutral: preresults are not results.

File: search/processor.cpp [hunk 7]
```diff
@@ -142,6 +142,7 @@ Processor::Processor(Index const & index, CategoriesHolder const & categories,
   : m_categories(categories)
   , m_infoGetter(infoGetter)
   , m_position(0, 0)
+  , m_minDistanceBetweenResults(0.0)
```
Comment (mpimenov) [inline search/processor.cpp]: s/Distance/DistanceOnMap/ ?

File: search/nearby_points_sweeper.cpp (hunk 2 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.cpp]: No, thanks. y is like a time (or priority) parameter. I can rename it, but I don't want to exchange it with x.

File: search/search_integration_tests/pre_ranker_test.cpp [hunk 8]
```diff
@@ -122,6 +113,7 @@ UNIT_CLASS_TEST(PreRankerTest, Smoke)
 
   PreRanker preRanker(m_engine, ranker, pois.size());
   PreRanker::Params params;
+  params.m_viewport = kViewport;
```
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: Shouldn't you set minDistanceBetweenResults here too?

File: search/pre_ranker.cpp (hunk 6 above)
Comment (ygorshenin) [inline search/pre_ranker.cpp]: We have PreRanker::m_results. So preResults here will require renaming of PreRanker::m_results too

File: map/framework.cpp (hunk 1 above)
Comment (ygorshenin) [inline map/framework.cpp]: Done

File: search/nearby_points_sweeper.hpp (hunk 3 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: Done

File: search/nearby_points_sweeper.hpp (hunk 4 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: Done

File: search/nearby_points_sweeper.hpp (hunk 5 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: Done

File: search/pre_ranker.hpp [hunk 9]
```diff
@@ -87,6 +95,11 @@ class PreRanker
   // Cache of nested rects used to estimate distance from a feature to the pivot.
   NestedRectsCache m_pivotFeatures;
 
+  set<FeatureID> m_currEmit;
```
Comment (mpimenov) [inline search/pre_ranker.hpp]: Add a comment to these variables, please.

File: search/processor.cpp (hunk 7 above)
Comment (ygorshenin) [inline search/processor.cpp]: Done

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 8 above)
Comment (ygorshenin) [inline search/search_integration_tests/pre_ranker_test.cpp]: We can, but does it make sense? All points here are distinct, and we don't want to sweep some of them.

File: search/pre_ranker.cpp (hunk 6 above)
Comment (ygorshenin) [inline search/pre_ranker.cpp]: Results of PreRanker are PreResults, results of Ranker are Results. Sorry, but I didn't get what's wrong here.

File: search/pre_ranker.hpp (hunk 9 above)
Comment (ygorshenin) [inline search/pre_ranker.hpp]: Done

**Comment by ygorshenin** [comment]
PTAL

File: search/pre_ranker.hpp [hunk 10]
```diff
@@ -87,6 +95,17 @@ class PreRanker
   // Cache of nested rects used to estimate distance from a feature to the pivot.
   NestedRectsCache m_pivotFeatures;
 
+  // A set of ids for features that are emit during the current search session.
```
Comment (mpimenov) [inline search/pre_ranker.hpp]: s/emit/emitted/
or
being emitted

File: search/nearby_points_sweeper.hpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+          line.emplace(event.m_x, event.m_index);
+          break;
+        }
+
+        auto it = line.upper_bound(make_pair(event.m_x, numeric_limits<size_t>::max()));
+
+        bool add = true;
+        while (true)
+        {
+          if (line.empty())
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: ```while (!line.empty())``` ?

File: search/nearby_points_sweeper.hpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+            break;
+
+          if (it == line.end())
+          {
+            --it;
+            continue;
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: no need.

File: search/nearby_points_sweeper.hpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+          }
+
+          double const x = it->first;
+          if (fabs(x - event.m_x) <= m_eps)
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: my::AlmostEqualAbs?

File: search/nearby_points_sweeper.hpp (hunk 12 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: I'd prefer to keep it as is. Here's a list of complex conditions, after all of them break or continue happens.

File: search/nearby_points_sweeper.hpp (hunk 11 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: No, thanks. This is not the main predicate for loop termination.

File: search/nearby_points_sweeper.hpp (hunk 13 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: Done

Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: Ahhh! No! AlmostEqualAbs checks that difference is less than eps, not less or equal.

**Comment by ygorshenin** [comment]
PTAL

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 8 above)
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: Set it to zero then.
There are currently two clients to this code: here and in Processor. While you set the radius explicitly in processor, here you rely on default initialization. Using this code for reference may result in an unexpected behaviour.

**Comment by rokuz** [comment]
LGTM

**Comment by ygorshenin** [comment]
@mpimenov PTAL

File: search/pre_ranker.hpp (hunk 10 above)
Comment (mpimenov) [inline search/pre_ranker.hpp]: ping on this comment

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 8 above)
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: ping on this comment

**Comment by mpimenov** [comment]
Change "nearby" to "viewport" in CL description?

File: search/nearby_points_sweeper.hpp [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+#include "std/algorithm.hpp"
+#include "std/cstdint.hpp"
+#include "std/limits.hpp"
+#include "std/set.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+namespace search
+{
+// This class can be used to sweep points on a plane that are too
```
Comment (mpimenov) [inline search/nearby_points_sweeper.hpp]: greedily sweep?

File: search/pre_ranker.hpp (hunk 10 above)
Comment (ygorshenin) [inline search/pre_ranker.hpp]: Done

File: search/nearby_points_sweeper.hpp (hunk 14 above)
Comment (ygorshenin) [inline search/nearby_points_sweeper.hpp]: Done

**Comment by ygorshenin** [comment]
PTAL

File: search/pre_ranker.hpp [hunk 15]
```diff
@@ -87,6 +95,17 @@ class PreRanker
   // Cache of nested rects used to estimate distance from a feature to the pivot.
   NestedRectsCache m_pivotFeatures;
 
+  // A set of ids for features that are emitted during the current search session.
+  set<FeatureID> m_currEmit;
+
+  // A set of ids for features that were emit during the previous
```
Comment (mpimenov) [inline search/pre_ranker.hpp]: s/emit/emitted/

Comment (ygorshenin) [inline search/pre_ranker.hpp]: Done

**Comment by ygorshenin** [comment]
PTAL

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**